### PR TITLE
[AArch64][SVE] Use INS when moving elements from bottom 128b of SVE type

### PR DESCRIPTION
**Pull request by SpencerAbson**
Moving elements from a scalable vector to a fixed-lengh vector should use[ INS (vector, element) ](https://developer.arm.com/documentation/100069/0606/SIMD-Vector-Instructions/INS--vector--element-) when we know that the extracted element is in the bottom 128-bits of the scalable vector. This avoids inserting unecessary UMOV/FMOV instructions.

### CODE REVIEW

**Comment by paulwalker-arm** [comment]
Is it possible to extend the existing `Neon_INS_elt_pattern` multi-class because it would be preferable to have the related logic in one place.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 1]
```diff
@@ -7222,44 +7222,101 @@ def : Pat<(v2i64 (int_aarch64_neon_vcopy_lane
...
+                                     ValueType VTScal, Operand ExIdxTy, Instruction INS> {
+  // Extracting from the lower 128-bits of an SVE vector
+  def : Pat<(VT128 (vector_insert VT128:$Rn,
+                      (VTScal (vector_extract VTSVE:$Rm, (i64 ExIdxTy:$Immn))),
+                      (i64 imm:$Immd))),
+            (INS VT128:$Rn, imm:$Immd, (VT128 (EXTRACT_SUBREG VTSVE:$Rm, zsub)), ExIdxTy:$Immn)>;
+
+  // Extracting from another Neon vector
+  def : Pat<(VT128 (vector_insert V128:$Rn,
+                        (VTScal (vector_extract (VT128 V128:$Rm), (i64 imm:$Immn))),
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:7236]: This is a much bigger change than I was expecting.  A better starting point would be to add the two new patterns to `Neon_INS_elt_pattern` that match the existing ones but take an SVE input.  I think that will implement the majority of the ISEL you need.

The remaining part is likely to be the two patterns needed for the `v1{i,f}64 scalar_to_vector` case, which I'd put in `AArch64InstrInfo.td` near the existing `sext_inreg (vector_extract` patterns.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 2]
```diff
@@ -7222,44 +7222,101 @@ def : Pat<(v2i64 (int_aarch64_neon_vcopy_lane
...
+                                ValueType VTScal, Operand ExIdxTy, Instruction INS> {
+  defm : Neon_INS_elt_pattern_v64<VT128, VT64, SVESrcVT, VTScal, ExIdxTy, INS>;
+  defm : Neon_INS_elt_pattern_v128<VT128, VT64, SVESrcVT, VTScal, ExIdxTy, INS>;
 }
 
-defm : Neon_INS_elt_pattern<v8f16, v4f16, f16, INSvi16lane>;
-defm : Neon_INS_elt_pattern<v8bf16, v4bf16, bf16, INSvi16lane>;
-defm : Neon_INS_elt_pattern<v4f32, v2f32, f32, INSvi32lane>;
-defm : Neon_INS_elt_pattern<v2f64, v1f64, f64, INSvi64lane>;
+defm : Neon_INS_elt_pattern<v4f32,  v2f32,  nxv4f32,  f32,  VectorIndexS, INSvi32lane>;
```
[the remark this answers is not in the document]
Comment (SpencerAbson) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td]: Thanks.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td (hunk 1 above)
Comment (SpencerAbson) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td]: Good point, this was a little over-kill.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 3]
```diff
@@ -7251,15 +7266,14 @@ multiclass Neon_INS_elt_pattern<ValueType VT128, ValueType VT64,
...
-defm : Neon_INS_elt_pattern<v4i32, v2i32, i32, INSvi32lane>;
-defm : Neon_INS_elt_pattern<v2i64, v1i64, i64, INSvi64lane>;
+defm : Neon_INS_elt_pattern<v8i16,  v4i16,  nxv8i16,  i32,  VectorIndexH, INSvi16lane>;
+defm : Neon_INS_elt_pattern<v8f16,  v4f16,  nxv8f16,  f16,  VectorIndexH, INSvi16lane>;
+defm : Neon_INS_elt_pattern<v8bf16, v4bf16, nxv8bf16, bf16, VectorIndexH, INSvi16lane>;
+defm : Neon_INS_elt_pattern<v16i8,  v8i8,   nxv16i8,  i32,  VectorIndexB, INSvi8lane>;
+defm : Neon_INS_elt_pattern<v4f32,  v2f32,  nxv4f32,  f32,  VectorIndexS, INSvi32lane>;
+defm : Neon_INS_elt_pattern<v4i32,  v2i32,  nxv4i32,  i32,  VectorIndexS, INSvi32lane>;
+defm : Neon_INS_elt_pattern<v2f64,  v1f64,  nxv2f64,  f64,  VectorIndexD, INSvi64lane>;
+defm : Neon_INS_elt_pattern<v2i64,  v1i64,  nxv2i64,  i64,  VectorIndexD, INSvi64lane>;
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:7276]: Sorry to nit pick but I don't see a good reason to reorder these?

**Review by paulwalker-arm — APPROVED** [review]

**Pull request merged**